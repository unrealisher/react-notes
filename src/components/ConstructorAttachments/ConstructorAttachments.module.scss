.attach {
  width: 100%;
}

.wrapper {
  display: flex;
  justify-content: space-between;
  margin-bottom: 12px;
}

.label {
  cursor: pointer;
  display: block;
  position: relative;
  width: 21px;
  height: 21px;
  border-radius: 50%;
  opacity: 0.4;

  &::before,
  &::after {
    content: "";
    position: absolute;
    width: 16px;
    height: 2px;
    top: calc(50% - 1px);
    left: calc(50% - 8px);
    background-color: green;
    transition: 0.6s;
  }

  &::after {
    transform: rotate(90deg);
  }

  &:hover {
    opacity: 1;
  }
}

.label_checked {
  &::before,
  &::after {
    transition: 0.6s;
    background-color: red;
  }

  &::before {
    transform: rotate(-45deg);
  }

  &::after {
    transform: rotate(45deg);
  }
}

.main {
  display: none;
}

.checkbox:checked + .main {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.margin_bottom {
  margin-bottom: 12px;
}

.list {
  display: flex;
  flex-wrap: wrap;
}

.item {
  position: relative;
  margin-right: 6px;
}

.label_file {
  width: 100%;
}

.file {
  box-sizing: border-box;
  width: 100%;
  padding: 2px 10px;
  border-radius: 15px;
}

.button {
  width: 100%;
  margin-top: 12px;
  padding: 5px 10px;
  border-radius: 15px;
  border: 1px solid rgba(218, 220, 224, 0.5);
  background-color: white;

  &:hover,
  &:focus {
    background-color: #fc0;
  }
}

.image {
  width: 80px;
  height: 80px;
  object-fit: cover;
}

.reset {
  cursor: pointer;
  position: absolute;
  top: -8px;
  left: calc(100% - 18px);
  width: 20px;
  height: 20px;
  background-color: rgb(124, 124, 124);
  border-radius: 50%;
  z-index: 2;
  opacity: 0.6;

  &::before,
  &::after {
    content: "";
    position: absolute;
    top: calc(50% - 1px);
    left: calc(50% - 7px);
    width: 14px;
    height: 2px;
    background-color: white;
  }

  &::before {
    transform: rotate(45deg);
  }

  &::after {
    transform: rotate(-45deg);
  }

  &:hover {
    opacity: 1;
  }

  &:active {
    opacity: 0.8;
  }
}
