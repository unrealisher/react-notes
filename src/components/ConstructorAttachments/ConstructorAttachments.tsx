import React, { useRef } from "react";
import cn from "classnames";

import TypeList from "../TypeList/TypeList";

import styles from "./ConstructorAttachments.module.scss";
import ListConstructor from "../ListConstructor/ListConstructor";

interface IProps {
  type: string;
  name: string;
  onTypeChange: Function;
  items: string[];
  setAttachItems: Function;
  checked: boolean;
  setChecked: Function;
}

export const ConstructorAttachments = (props: IProps): JSX.Element => {
  const {
    type,
    name,
    onTypeChange,
    items,
    setAttachItems,
    checked,
    setChecked
  } = props;
  const inputRef = useRef<HTMLInputElement>(null);

  const getImages = (): JSX.Element => {
    return (
      <ul className={styles.list}>
        {items.map(item => {
          return (
            <li key={item} className={styles.item}>
              <img
                src={item}
                className={styles.image}
                alt="Что-то пошло не так"
              />
              <button
                className={styles.reset}
                onClick={() => {
                  let arr = items.slice();
                  const index = items.indexOf(item);
                  if (index !== -1) arr.splice(index, 1);
                  setAttachItems(arr);
                }}
              />
            </li>
          );
        })}
      </ul>
    );
  };

  return (
    <div className={styles.attach}>
      <div className={styles.wrapper}>
        <span className={styles.title}>Дополнения</span>
        <label
          htmlFor="checkbox"
          className={cn(styles.label, checked && styles.label_checked)}
        />
      </div>
      <input
        type="checkbox"
        className={styles.checkbox}
        checked={checked}
        onChange={() => setChecked(!checked)}
        id="checkbox"
        hidden
      />
      <div className={styles.main}>
        <TypeList
          wrapper={styles.margin_bottom}
          type={type}
          name={name}
          onTypeChange={onTypeChange}
          items={[
            { text: "Ссылка", value: "link" },
            { text: "Картинка", value: "image" }
          ]}
        />
        {type === "link" && (
          <ListConstructor
            wrapper={styles.margin_bottom}
            items={items}
            setItems={setAttachItems}
          />
        )}
        {type === "image" && (
          <>
            {items.length > 0 && getImages()}
            <label className={styles.label_file}>
              Введите URL картинки:
              <input className={styles.file} type="text" ref={inputRef} />
            </label>
            <button
              className={styles.button}
              onClick={() => {
                let arr = items.slice();
                if (inputRef.current) {
                  arr.push(inputRef.current.value);
                  inputRef.current.value = "";
                }
                setAttachItems(arr);
              }}
            >
              Загрузить картинку
            </button>
          </>
        )}
      </div>
    </div>
  );
};

export default ConstructorAttachments;
