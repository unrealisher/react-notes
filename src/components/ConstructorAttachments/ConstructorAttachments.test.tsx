import React from "react";
import { shallow, mount } from "enzyme";
import { ConstructorAttachments } from "./ConstructorAttachments";

describe("ConstructorAttachments", () => {
  const type = "link";
  const name = "link";
  const onTypeChange = (): void => {};
  const items = [
    "https://avatars.mds.yandex.net/get-pdb/1816426/93eab951-b130-4cf9-98d6-01e250be5530/orig",
    "https://avatars.mds.yandex.net/get-pdb/1823123/fc6cf893-91b5-4e80-9f02-82acfc6fe30c/orig",
    "https://avatars.mds.yandex.net/get-pdb/1816426/c68f1a18-f763-4a98-8a17-f7d7725753b6/orig"
  ];
  const setAttachItems = (): void => {};
  const checked = true;
  const setChecked = (): void => {};

  it("correct render", () => {
    const component = shallow(
      <ConstructorAttachments
        type={type}
        name={name}
        onTypeChange={onTypeChange}
        items={items}
        setAttachItems={setAttachItems}
        checked={checked}
        setChecked={setChecked}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("call setChecked when checkbox checked", () => {
    const setChecked = jest.fn();
    const component = mount(
      <ConstructorAttachments
        type={type}
        name={name}
        onTypeChange={onTypeChange}
        items={items}
        setAttachItems={setAttachItems}
        checked={checked}
        setChecked={setChecked}
      />
    );
    const checkbox = component.find(".checkbox");
    checkbox.simulate("change");
    expect(setChecked).toBeCalledWith(!checked);
  });
});
