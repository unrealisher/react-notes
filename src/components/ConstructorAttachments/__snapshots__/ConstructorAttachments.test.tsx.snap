// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConstructorAttachments correct render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ConstructorAttachments
    checked={true}
    items={
      Array [
        "https://avatars.mds.yandex.net/get-pdb/1816426/93eab951-b130-4cf9-98d6-01e250be5530/orig",
        "https://avatars.mds.yandex.net/get-pdb/1823123/fc6cf893-91b5-4e80-9f02-82acfc6fe30c/orig",
        "https://avatars.mds.yandex.net/get-pdb/1816426/c68f1a18-f763-4a98-8a17-f7d7725753b6/orig",
      ]
    }
    name="link"
    onTypeChange={[Function]}
    setAttachItems={[Function]}
    setChecked={[Function]}
    type="link"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="wrapper"
        >
          <span
            className="title"
          >
            Дополнения
          </span>
          <label
            className="label label_checked"
            htmlFor="checkbox"
          />
        </div>,
        <input
          checked={true}
          className="checkbox"
          hidden={true}
          id="checkbox"
          onChange={[Function]}
          type="checkbox"
        />,
        <div
          className="main"
        >
          <TypeList
            items={
              Array [
                Object {
                  "text": "Ссылка",
                  "value": "link",
                },
                Object {
                  "text": "Картинка",
                  "value": "image",
                },
              ]
            }
            name="link"
            onTypeChange={[Function]}
            type="link"
            wrapper="margin_bottom"
          />
          <ListConstructor
            items={
              Array [
                "https://avatars.mds.yandex.net/get-pdb/1816426/93eab951-b130-4cf9-98d6-01e250be5530/orig",
                "https://avatars.mds.yandex.net/get-pdb/1823123/fc6cf893-91b5-4e80-9f02-82acfc6fe30c/orig",
                "https://avatars.mds.yandex.net/get-pdb/1816426/c68f1a18-f763-4a98-8a17-f7d7725753b6/orig",
              ]
            }
            setItems={[Function]}
            wrapper="margin_bottom"
          />
        </div>,
      ],
      "className": "attach",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="title"
            >
              Дополнения
            </span>,
            <label
              className="label label_checked"
              htmlFor="checkbox"
            />,
          ],
          "className": "wrapper",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Дополнения",
              "className": "title",
            },
            "ref": null,
            "rendered": "Дополнения",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "label label_checked",
              "htmlFor": "checkbox",
            },
            "ref": null,
            "rendered": null,
            "type": "label",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "checked": true,
          "className": "checkbox",
          "hidden": true,
          "id": "checkbox",
          "onChange": [Function],
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TypeList
              items={
                Array [
                  Object {
                    "text": "Ссылка",
                    "value": "link",
                  },
                  Object {
                    "text": "Картинка",
                    "value": "image",
                  },
                ]
              }
              name="link"
              onTypeChange={[Function]}
              type="link"
              wrapper="margin_bottom"
            />,
            <ListConstructor
              items={
                Array [
                  "https://avatars.mds.yandex.net/get-pdb/1816426/93eab951-b130-4cf9-98d6-01e250be5530/orig",
                  "https://avatars.mds.yandex.net/get-pdb/1823123/fc6cf893-91b5-4e80-9f02-82acfc6fe30c/orig",
                  "https://avatars.mds.yandex.net/get-pdb/1816426/c68f1a18-f763-4a98-8a17-f7d7725753b6/orig",
                ]
              }
              setItems={[Function]}
              wrapper="margin_bottom"
            />,
            false,
          ],
          "className": "main",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "items": Array [
                Object {
                  "text": "Ссылка",
                  "value": "link",
                },
                Object {
                  "text": "Картинка",
                  "value": "image",
                },
              ],
              "name": "link",
              "onTypeChange": [Function],
              "type": "link",
              "wrapper": "margin_bottom",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "items": Array [
                "https://avatars.mds.yandex.net/get-pdb/1816426/93eab951-b130-4cf9-98d6-01e250be5530/orig",
                "https://avatars.mds.yandex.net/get-pdb/1823123/fc6cf893-91b5-4e80-9f02-82acfc6fe30c/orig",
                "https://avatars.mds.yandex.net/get-pdb/1816426/c68f1a18-f763-4a98-8a17-f7d7725753b6/orig",
              ],
              "setItems": [Function],
              "wrapper": "margin_bottom",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="wrapper"
          >
            <span
              className="title"
            >
              Дополнения
            </span>
            <label
              className="label label_checked"
              htmlFor="checkbox"
            />
          </div>,
          <input
            checked={true}
            className="checkbox"
            hidden={true}
            id="checkbox"
            onChange={[Function]}
            type="checkbox"
          />,
          <div
            className="main"
          >
            <TypeList
              items={
                Array [
                  Object {
                    "text": "Ссылка",
                    "value": "link",
                  },
                  Object {
                    "text": "Картинка",
                    "value": "image",
                  },
                ]
              }
              name="link"
              onTypeChange={[Function]}
              type="link"
              wrapper="margin_bottom"
            />
            <ListConstructor
              items={
                Array [
                  "https://avatars.mds.yandex.net/get-pdb/1816426/93eab951-b130-4cf9-98d6-01e250be5530/orig",
                  "https://avatars.mds.yandex.net/get-pdb/1823123/fc6cf893-91b5-4e80-9f02-82acfc6fe30c/orig",
                  "https://avatars.mds.yandex.net/get-pdb/1816426/c68f1a18-f763-4a98-8a17-f7d7725753b6/orig",
                ]
              }
              setItems={[Function]}
              wrapper="margin_bottom"
            />
          </div>,
        ],
        "className": "attach",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="title"
              >
                Дополнения
              </span>,
              <label
                className="label label_checked"
                htmlFor="checkbox"
              />,
            ],
            "className": "wrapper",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Дополнения",
                "className": "title",
              },
              "ref": null,
              "rendered": "Дополнения",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "label label_checked",
                "htmlFor": "checkbox",
              },
              "ref": null,
              "rendered": null,
              "type": "label",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": true,
            "className": "checkbox",
            "hidden": true,
            "id": "checkbox",
            "onChange": [Function],
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <TypeList
                items={
                  Array [
                    Object {
                      "text": "Ссылка",
                      "value": "link",
                    },
                    Object {
                      "text": "Картинка",
                      "value": "image",
                    },
                  ]
                }
                name="link"
                onTypeChange={[Function]}
                type="link"
                wrapper="margin_bottom"
              />,
              <ListConstructor
                items={
                  Array [
                    "https://avatars.mds.yandex.net/get-pdb/1816426/93eab951-b130-4cf9-98d6-01e250be5530/orig",
                    "https://avatars.mds.yandex.net/get-pdb/1823123/fc6cf893-91b5-4e80-9f02-82acfc6fe30c/orig",
                    "https://avatars.mds.yandex.net/get-pdb/1816426/c68f1a18-f763-4a98-8a17-f7d7725753b6/orig",
                  ]
                }
                setItems={[Function]}
                wrapper="margin_bottom"
              />,
              false,
            ],
            "className": "main",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "items": Array [
                  Object {
                    "text": "Ссылка",
                    "value": "link",
                  },
                  Object {
                    "text": "Картинка",
                    "value": "image",
                  },
                ],
                "name": "link",
                "onTypeChange": [Function],
                "type": "link",
                "wrapper": "margin_bottom",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "items": Array [
                  "https://avatars.mds.yandex.net/get-pdb/1816426/93eab951-b130-4cf9-98d6-01e250be5530/orig",
                  "https://avatars.mds.yandex.net/get-pdb/1823123/fc6cf893-91b5-4e80-9f02-82acfc6fe30c/orig",
                  "https://avatars.mds.yandex.net/get-pdb/1816426/c68f1a18-f763-4a98-8a17-f7d7725753b6/orig",
                ],
                "setItems": [Function],
                "wrapper": "margin_bottom",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
